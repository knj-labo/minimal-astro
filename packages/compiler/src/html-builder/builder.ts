/**
 * AST to HTML builder
 *
 * This module provides the main entry point for converting
 * an Astro AST into an HTML string.
 *
 * @module builder
 */

import type { AstroAST, AstroNode } from '../parser/ast.js'
import { serializeElement, serializeText } from './serializer.js'

/**
 * Set of HTML void elements that are self-closing
 */
const VOID_ELEMENTS = new Set([
  'img',
  'br',
  'hr',
  'input',
  'area',
  'base',
  'col',
  'embed',
  'link',
  'meta',
  'source',
  'track',
  'wbr',
])

/**
 * Checks if an element is a void element
 */
function isVoidElement(tagName: string): boolean {
  return VOID_ELEMENTS.has(tagName.toLowerCase())
}

/**
 * Builds an HTML string from an AST node
 *
 * Recursively processes AST nodes and converts them to HTML.
 * Handles different node types appropriately:
 * - Text nodes are escaped for safe HTML rendering
 * - Element nodes are serialized with attributes and children
 * - Container nodes (Program, Template) concatenate their children
 * - Other nodes (Frontmatter, Expression) are skipped or handled specially
 *
 * @param node - The AST node to convert
 * @returns HTML string representation
 */
function buildNode(node: AstroNode): string {
  switch (node.type) {
    case 'Text':
      return serializeText(node)

    case 'Element':
      // Check if it's a void element and update selfClosing accordingly
      if (isVoidElement(node.name)) {
        return serializeElement({ ...node, selfClosing: true })
      }
      return serializeElement(node)

    case 'Expression':
      // Expressions in the template should be rendered as-is for now
      // In a full implementation, these would be evaluated
      return `{${node.value}}`

    case 'Program':
    case 'Template':
      // Container nodes: concatenate all children
      return node.children.map(child => buildNode(child)).join('')

    case 'Frontmatter':
      // Frontmatter is not rendered in the HTML output
      return ''

    default:
      // Unknown node types are ignored
      return ''
  }
}

/**
 * Builds an HTML string from an Astro AST
 *
 * This is the main entry point for converting a parsed Astro AST
 * into HTML. It processes the entire AST tree and returns the
 * resulting HTML string.
 *
 * @param ast - The Astro AST to convert
 * @returns Complete HTML string
 *
 * @example
 * ```typescript
 * const ast = parse('<p>Hello World</p>')
 * const html = buildHTML(ast)
 * console.log(html) // '<p>Hello World</p>'
 * ```
 */
export function buildHTML(ast: AstroAST): string {
  if (!ast || !ast.children || ast.children.length === 0) {
    return ''
  }

  return ast.children.map(child => buildNode(child)).join('')
}
