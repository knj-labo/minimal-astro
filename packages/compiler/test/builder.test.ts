import { describe, expect, it } from 'vitest'
import { buildHTML } from '../src/html-builder/builder.js'
import type { AstroAST } from '../src/parser/ast.js'

describe('buildHTML', () => {
  it('空のAST → ""を返す', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [],
    }
    expect(buildHTML(ast)).toBe('')
  })

  it('単一のテキスト → "Hello"を返す', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Text',
          value: 'Hello',
        },
      ],
    }
    expect(buildHTML(ast)).toBe('Hello')
  })

  it('単一の要素 → <p>Hello</p>', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Element',
          name: 'p',
          attributes: [],
          children: [
            {
              type: 'Text',
              value: 'Hello',
            },
          ],
          selfClosing: false,
        },
      ],
    }
    expect(buildHTML(ast)).toBe('<p>Hello</p>')
  })

  it('ネストされた要素 → <div><span>x</span></div>', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Element',
          name: 'div',
          attributes: [],
          children: [
            {
              type: 'Element',
              name: 'span',
              attributes: [],
              children: [
                {
                  type: 'Text',
                  value: 'x',
                },
              ],
              selfClosing: false,
            },
          ],
          selfClosing: false,
        },
      ],
    }
    expect(buildHTML(ast)).toBe('<div><span>x</span></div>')
  })

  it('自己完結型要素 → <img src="a.png">', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Element',
          name: 'img',
          attributes: [{ name: 'src', value: 'a.png' }],
          children: [],
          selfClosing: false, // Even if false, img should be self-closing
        },
      ],
    }
    expect(buildHTML(ast)).toBe('<img src="a.png" />')
  })

  it('フロントマターを含むASTはフロントマターを無視する', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Frontmatter',
          value: 'const title = "Hello"',
        },
        {
          type: 'Element',
          name: 'h1',
          attributes: [],
          children: [
            {
              type: 'Text',
              value: 'Hello World',
            },
          ],
          selfClosing: false,
        },
      ],
    }
    expect(buildHTML(ast)).toBe('<h1>Hello World</h1>')
  })

  it('Expressionノードを含むAST', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Element',
          name: 'div',
          attributes: [],
          children: [
            {
              type: 'Text',
              value: 'Count: ',
            },
            {
              type: 'Expression',
              value: 'count',
            },
          ],
          selfClosing: false,
        },
      ],
    }
    expect(buildHTML(ast)).toBe('<div>Count: {count}</div>')
  })

  it('Templateノードを含むAST', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Template',
          children: [
            {
              type: 'Element',
              name: 'main',
              attributes: [],
              children: [
                {
                  type: 'Text',
                  value: 'Content',
                },
              ],
              selfClosing: false,
            },
          ],
        },
      ],
    }
    expect(buildHTML(ast)).toBe('<main>Content</main>')
  })

  it('複雑なASTのスナップショット', () => {
    const ast: AstroAST = {
      type: 'Program',
      children: [
        {
          type: 'Frontmatter',
          value: 'const title = "My Page"',
        },
        {
          type: 'Element',
          name: 'html',
          attributes: [{ name: 'lang', value: 'ja' }],
          children: [
            {
              type: 'Element',
              name: 'head',
              attributes: [],
              children: [
                {
                  type: 'Element',
                  name: 'title',
                  attributes: [],
                  children: [
                    {
                      type: 'Expression',
                      value: 'title',
                    },
                  ],
                  selfClosing: false,
                },
                {
                  type: 'Element',
                  name: 'meta',
                  attributes: [{ name: 'charset', value: 'UTF-8' }],
                  children: [],
                  selfClosing: true,
                },
              ],
              selfClosing: false,
            },
            {
              type: 'Element',
              name: 'body',
              attributes: [],
              children: [
                {
                  type: 'Element',
                  name: 'h1',
                  attributes: [{ name: 'class', value: 'heading' }],
                  children: [
                    {
                      type: 'Text',
                      value: 'Welcome to ',
                    },
                    {
                      type: 'Expression',
                      value: 'title',
                    },
                  ],
                  selfClosing: false,
                },
                {
                  type: 'Element',
                  name: 'p',
                  attributes: [],
                  children: [
                    {
                      type: 'Text',
                      value: 'This is a <test> of HTML escaping & special "characters".',
                    },
                  ],
                  selfClosing: false,
                },
                {
                  type: 'Element',
                  name: 'img',
                  attributes: [
                    { name: 'src', value: '/image.png' },
                    { name: 'alt', value: 'Test Image' },
                  ],
                  children: [],
                  selfClosing: false,
                },
              ],
              selfClosing: false,
            },
          ],
          selfClosing: false,
        },
      ],
    }

    expect(buildHTML(ast)).toMatchSnapshot()
  })
})
