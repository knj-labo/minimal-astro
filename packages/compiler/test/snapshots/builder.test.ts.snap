// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildHTML > 複雑なASTのスナップショット 1`] = `"<html lang="ja"><head><title>{title}</title><meta charset="UTF-8" /></head><body><h1 class="heading">Welcome to {title}</h1><p>This is a &lt;test&gt; of HTML escaping &amp; special &quot;characters&quot;.</p><img src="/image.png" alt="Test Image"></img></body></html>"`;
